packages/SwingSet/src/liveslots/virtualReferences.js

```diff
@@ -466,43 +466,56 @@ export function makeVirtualReferenceManager(
   /** @type {Map<string, Set<Recognizer>>} */
   const vrefRecognizers = new Map();
 
-  function addRecognizableValue(value, recognizer) {
+  function addRecognizableValue(value, recognizer, recognizerIsVirtual) {
     const vref = getSlotForVal(value);
     if (vref) {
       const { type, allocatedByVat, virtual } = parseVatSlot(vref);
       if (type === 'object' && (!allocatedByVat || virtual)) {
-        let recognizerSet = vrefRecognizers.get(vref);
-        if (!recognizerSet) {
-          recognizerSet = new Set();
-          vrefRecognizers.set(vref, recognizerSet);
+        if (recognizerIsVirtual) {
+          syscall.vatstoreSet(`vom.ir.${vref}|${recognizer}`, '1');
+        } else {
+          let recognizerSet = vrefRecognizers.get(vref);
+          if (!recognizerSet) {
+            recognizerSet = new Set();
+            vrefRecognizers.set(vref, recognizerSet);
+          }
+          recognizerSet.add(recognizer);
         }
-        recognizerSet.add(recognizer);
       }
     }
   }
 
-  function removeRecognizableVref(vref, recognizer) {
+  function removeRecognizableVref(vref, recognizer, recognizerIsVirtual) {
     const { type, allocatedByVat, virtual } = parseVatSlot(vref);
     if (type === 'object' && (!allocatedByVat || virtual)) {
-      const recognizerSet = vrefRecognizers.get(vref);
-      assert(recognizerSet && recognizerSet.has(recognizer));
-      recognizerSet.delete(recognizer);
-      if (recognizerSet.size === 0) {
-        vrefRecognizers.delete(vref);
-        if (!allocatedByVat) {
-          addToPossiblyRetiredSet(vref);
+      if (recognizerIsVirtual) {
+        syscall.vatstoreDelete(`vom.ir.${vref}|${recognizer}`);
```

Comment:
This introduces a constraint that vrefs cannot have `|` in them, right? We don't currently do that, but we should make a note of it, so we don't accidentally allow them in the future.

* add a brief section to your excellent comment in `parseVatSlot()` about characters which are prohibited from vrefs because other tables need them for delimiters, and give it a one-entry list naming `|` and a pointer to `removeRecognizableVref` and the `vom.ir.${vref}|${recognizer}` keys it uses
* Also update `src/liveslots/vatstore-usage.md` with an explanation of the new `vom.ir.*` keys

Reply:
I believe we already had the exclusion of `|` due to its use as a delimiter elsewhere, but even if that recollection is faulty you're right that this exclusion should be documented.  I have updated `parseVatSlot` and the docs accordingly.